Sources/PackageCollectionsSigning/Certificate/CertificatePolicy.swift

```diff
@@ -82,19 +86,295 @@ extension CertificatePolicy {
+            // Success
+            if result == 1 { return result }
+
+            // Custom error handling
+            let errorCode = CCryptoBoringSSL_X509_STORE_CTX_get_error(ctx)
+            // Certs could have unknown critical extensions and cause them to be rejected.
+            // Instead of disabling all critical extension checks with X509_V_FLAG_IGNORE_CRITICAL
+            // we will just ignore this specific error.
+            if errorCode == X509_V_ERR_UNHANDLED_CRITICAL_EXTENSION {
+                return 1
+            }
+            return result
+        }
+        CCryptoBoringSSL_X509_STORE_CTX_set_verify_cb(x509StoreCtx, verifyCallback)
+
+        guard CCryptoBoringSSL_X509_verify_cert(x509StoreCtx) == 1 else {
+            let error = CCryptoBoringSSL_X509_verify_cert_error_string(numericCast(CCryptoBoringSSL_X509_STORE_CTX_get_error(x509StoreCtx)))
+            diagnosticsEngine.emit(warning: "The certificate is invalid: \(String(describing: error))")
+            return wrappedCallback(.failure(CertificatePolicyError.invalidCertChain))
+        }
+
+        if certChain.count >= 1, let httpClient = httpClient {
+            // Whether cert chain can be trusted depends on OCSP result
+            ocspClient.checkStatus(certificate: certChain[0], issuer: certChain[1], httpClient: httpClient,
+                                   diagnosticsEngine: diagnosticsEngine, callbackQueue: callbackQueue, callback: callback)
+        } else {
+            wrappedCallback(.success(()))
+        }
     }
+    #endif
 }
 
+#if os(Linux) || os(Windows)
+private let ocspClient = BoringSSLOCSPClient()
+
+private struct BoringSSLOCSPClient {
+    private let resultCache = ThreadSafeKeyValueStore<CacheKey, CacheValue>()
+
+    private let cacheTTL: DispatchTimeInterval
+
+    init(cacheTTL: DispatchTimeInterval = .seconds(300)) {
+        self.cacheTTL = cacheTTL
+    }
+
+    func checkStatus(certificate: Certificate,
+                     issuer: Certificate,
+                     httpClient: HTTPClient,
+                     diagnosticsEngine: DiagnosticsEngine,
+                     callbackQueue: DispatchQueue,
+                     callback: @escaping (Result<Void, Error>) -> Void) {
+        let wrappedCallback: (Result<Void, Error>) -> Void = { result in callbackQueue.async { callback(result) } }
+
+        let ocspURLs = CCryptoBoringSSL_X509_get1_ocsp(certificate.underlying)
+        defer { CCryptoBoringSSL_sk_OPENSSL_STRING_free(ocspURLs) }
+
+        let ocspURLCount = CCryptoBoringSSL_sk_OPENSSL_STRING_num(ocspURLs)
+        // Nothing to do if no OCSP URLs. Use `supportsOCSP` to require OCSP support if needed.
+        guard ocspURLCount > 0 else { return wrappedCallback(.success(())) }
+
+        // Construct the OCSP request
+        let digest = CCryptoBoringSSL_EVP_sha1()
+        guard let certid = OCSP_cert_to_id(digest, certificate.underlying, issuer.underlying),
```

Comment:
The lifetimes of `certificate.underlying` and `issuer.underlying` are strictly bounded by that of `certificate` and `issuer`. The pointers can be left dangling here if we're not careful. Either `certificate` and `issuer` should only provide their interior pointers by way of `with` functions or we need to use `withExtendedLifetime` to extend the lifetime of both `certificate` and `issuer`.

Reply:
Added `withUnsafeMutablePointer` to `Certificate` https://github.com/apple/swift-package-manager/pull/3272/commits/b844e070ad3dbf654115133ba31c97cb164f0128